### ISSUE TITLE
False positive CS4007 "Instance of type 'System.ReadOnlySpan<Buffer10<int>>' cannot be preserved across 'await' or 'yield' boundary"

### ISSUE THREAD

**Issue by sirduke** [dotnet/roslyn#74046]
**Version Used**:  [9d5e37502942b6a727daffc74fa2327ed22f22e6](https://github.com/dotnet/roslyn/commit/9d5e37502942b6a727daffc74fa2327ed22f22e6)

**Steps to Reproduce**:
```

using System.Threading.Tasks;
using System;

class Program22
{
    static async Task<int> M1()
    {
        ReadOnlySpan<Buffer10<int>> readOnlySpan = default;
        return readOnlySpan[0][await Task.FromResult(1)]; // error "Instance of type 'System.ReadOnlySpan<Buffer10<int>>' cannot be preserved across 'await' or 'yield' boundary"
    }
}

[System.Runtime.CompilerServices.InlineArray(10)]
public struct Buffer10<T>
{
    private T _element0;
}
```

[Sample in sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIAMABHgIwCsA3ALABQeATKXXY0QAoBOA9gOacQwDBnQDedIpJKkAbCQAcJGQB5YCAHxEAsqQAUASglTxtKWaIAlAKYR0AeRgoAngGUADhBjKAQgFcAZv5WnKQEqjAampw29o6uHjBEALxE6Fb+EL4oCDSm5pJ4AOxE0bYOzu6eANoEALpVeACcSgB0AGI8YNYAzlkIuqT6tZREAPSjRME8nEQARACSMN0IngDGVkTc/kQITm4bAORkLdZlcZVefoHBoeGRB0SrnjDcCETAG27R3cEAblboIgQVY8brdIhHRoPbgzI6kUgPYDcXwwdAQThOWZGSQAX2xRHx+LIcjwijwKjUmi0DAM+JM+Ukp1iFQSPgCQRCYUpURi5XinmSqXSmWyuQZRF+6Pe7JuxBSpWZ/JgNWG+LMRWl105DWa5PanR6fQGQxG4yIL0mnGm+LxtFtdAapBOKIQUDAVhaAGFuGA3Khgi4/lB1t0WosULArABBK0QJwDAhDVgAZiIy04vlWbyuHNuABV1GJ8V8oJKEBs80QAPpWFBWd0RAi5HFAA===)

**Diagnostic Id**:

CS4007 "Instance of type 'System.ReadOnlySpan<Buffer10<int>>' cannot be preserved across 'await' or 'yield' boundary"

**Expected Behavior**:
No error, code is compilable

**Actual Behavior**:
Error CS4007 "Instance of type 'System.ReadOnlySpan<Buffer10<int>>' cannot be preserved across 'await' or 'yield' boundary"

**Comment by jjonescz** [comment]
I think the error is expected because the ReadOnlySpan needs to be on stack when invoking the Buffer10's indexer (which happens after the `await`). The return expression in your example lowers to:

```cs
return <PrivateImplementationDetails>.InlineArrayAsReadOnlySpan<Buffer10<int>, int>(ref readonlySpan[0], 10)[await Task.FromResult(1)];
```

**Comment by jaredpar** [comment]
As noted this behavior is "By Design".
